"Global" state - preparation

gomorizsolt opened the pull request:
TODO(s):
- [x] merge PR #11 into master,
- [x] refinements.

---

Part of issue #12.

The PR is a preparation for the further changes. Basically, the results are as follows:
- creates the "global" `state` object which:
   - stores the received configs via the `constructor`(we can even extend this with the other expected values - GH / GL users - later, thereby the passed params towards the functions are abbreviated),
   - sets the "state" immutably,
   - holds the logic of calendar rendering.
- a restructured, cleaner(?) flow.

As it can be seen, the state object roughly behaves in the same way that we experience with React.

thisismydesign commented on src/Main/Main.js:
```diff
@@ -1,5 +1,10 @@
-  await CalendarUtils.initializeBasicAppearance(container, proxyServerUrl);
+  const state = new State(container, proxyServerUrl);
+
+  state.render();
```
If we'll call this state it shouldn't know how to render. So it's either something else, or (and i like the separation here) we have something that can render a state. (which is already the case actually, it's just used from the state)

gomorizsolt replied:
Regarding this proposal, could you check my [**comment**](https://github.com/c-hive/team-contribution-calendar/pull/13#discussion_r276108299) below?

thisismydesign commented on src/utils/CalendarUtils/Render/Render.test.js:
```diff
@@ -52,15 +56,62 @@ describe('Render', () => {
+      )).to.equal(true);
+    });
+
+    it('sets the restored user calendar to the state and increments the total contributions by 0', async () => {
```
Let's name this according to the call: `calls state.setStateAndRender with ...`

thisismydesign commented on src/utils/CalendarUtils/Render/Render.js:
```diff
@@ -10,3 +12,16 @@ export const calendarWithContributions = (container, calendar, totalContribution
+
+  const restoredDefaultUserCalendar = GitHub.restoreCalendarValues(defaultUserJsonCalendar);
+
+  state.setStateAndRender({
```
State**And**Render sounds like an indication that these could be separated :)

gomorizsolt replied:
Actually, the `render` is separated, but you're right.

Btw, my concern is that after each `setState` call we'd like to re-render the whole calendar. As a result, the same lines of codes will be written similarly to this example:

```
state.setState({
   updatedActualCalendar: BasicCalendar,
   currentUserTotalContributions: 0,
});

CalendarUtils.Render.calendarWithContributions(state);
```

I suspect that it's an indication to reconsider the current structure. How about creating a `Calendar`/`CalendarDetails` - instead of `State` as the naming can be misleading - class that holds:
- the whole logic of rendering the actual calendar according to the state of the fields within the class - and also able to establish the basic appearance(basic calendar, default user calendar),
- after each data-update, it re-renders the screen by calling its own renderer.

Example:
```
export default class Calendar {
   constructor(...) {
      // receives the "configs" and sets the default values as before
   }

   renderActualCalendar() {
      // ...
   }

   async renderBasicAppearance() {
      // renders BasicCalendar and the default user calendar synchronously
   }

   updateCalendarDetails(data) {
      // update values, call renderActualCalendar - or run it manually after each data-update
   }
};
```

It can even expose other functions later.

thisismydesign replied:
I like that, but i'd still keep the State separately and use accessors.

thisismydesign commented on src/utils/CalendarUtils/Render/Render.js:
```diff
@@ -10,3 +12,16 @@ export const calendarWithContributions = (container, calendar, totalContribution
+    state.configs.proxyServerUrl, DefaultUsers.GitHub,
+  );
+
+  const restoredDefaultUserCalendar = GitHub.restoreCalendarValues(defaultUserJsonCalendar);
```
`restoreCalendarValues` implies some previous state. How about calling it `setEmptyCalendarValues`?

gomorizsolt commented:
Recent changes(based on the flow that I mentioned above/on Slack):
- removed the `Render` folder,
- created `TeamContributionCalendar`(the previously deleted functions are placed here),
- updated the tests that are related to the modifications.
